h2o-extensions/target-encoder/src/main/java/ai/h2o/targetencoding/TargetEncoderModel.java

```diff
@@ -84,48 +85,79 @@ protected boolean defaultDropConsCols() {
 
     public final TargetEncoderParameters _parms;
     public final int _nclasses;
+    public final ColumnsToSingleMapping[] _input_to_encoding_column; // maps input columns (or groups of columns) to the single column being effectively encoded (= key in _target_encoding_map).
+    public final ColumnsMapping[] _input_to_output_columns; // maps input columns (or groups of columns) to their corresponding encoded one(s).
     public final IcedHashMap<String, Frame> _target_encoding_map;
     public final IcedHashMap<String, Boolean> _te_column_to_hasNAs; //XXX: Map is a wrong choice for this, IcedHashSet would be perfect though
     
-    public TargetEncoderOutput(TargetEncoder b, IcedHashMap<String, Frame> teMap) {
-      super(b);
-      _parms = b._parms;
-      _nclasses = b.nclasses();
+    
+    public TargetEncoderOutput(TargetEncoder te) {
+      this(te, new IcedHashMap<>(), new ColumnsToSingleMapping[]{});
+    }
+    
+    public TargetEncoderOutput(TargetEncoder te, IcedHashMap<String, Frame> teMap, ColumnsToSingleMapping[] columnsToEncodeMapping) {
```

Comment:
Ok, I see what you  mean by `Optional` parameter now.
However, in Java, as much as I see the use for `Optional` in return values (I think that `Optional` was inspired from FP languages like Haskell that has this concept of `Maybe` return values to be able to chain functions), I don't see the benefit as a parameter. Actually I think the good old method overload in Java is much more clear at saying what is optional or not and I've never seen `Optional` used this way before in JDK or famous libraries.

Reply:
From B. Goetz himself (yes, a true authority argument :), but that's before all because I agree), (https://stackoverflow.com/questions/26327957/should-java-8-getters-return-optional-type/26328555#26328555):

---
Of course, people will do what they want. But we did have a clear intention when adding this feature, and it was not to be a general purpose Maybe type, as much as many people would have liked us to do so. Our intention was to provide a limited mechanism for library method return types where there needed to be a clear way to represent "no result", and using null for such was overwhelmingly likely to cause errors.

For example, you probably should never use it for something that returns an array of results, or a list of results; instead return an empty array or list. **You should almost never use it as a field of something or a method parameter.**

**There's nothing wrong with Optional that it should be avoided, it's just not what many people wish it were, and accordingly we were fairly concerned about the risk of zealous over-use.**

Reply:
Sure, that's what I meant. It's up to you to decide. I'm not pushing it. Not a deal breaker. Deal breaker would be big inefficiency or bad business logic. :wink: 

Monkey see, monkey writes - when I see something during an interview, I point it out. Not everything is a must.